monkey/infection_monkey/network/network_scanner.py

```diff
@@ -68,49 +97,39 @@ def get_victim_machines(self, max_find=5, stop_callback=None):
         :param stop_callback: A callback to check at any point if we should stop scanning
         :return: yields a sequence of VictimHost instances
         """
-
-        TCPscan = TcpScanner()
-        Pinger = PingScanner()
+        pool = Pool()
         victims_count = 0
+        for victim_chunk in generate_victims(self._ranges, ITERATION_BLOCK_SIZE):
+            LOG.debug("Scanning for potential victims in chunk %r", victim_chunk)
 
-        for net_range in self._ranges:
-            LOG.debug("Scanning for potential victims in the network %r", net_range)
-            for ip_addr in net_range:
-                if hasattr(net_range, 'domain_name'):
-                    victim = VictimHost(ip_addr, net_range.domain_name)
-                else:
-                    victim = VictimHost(ip_addr)
-                if stop_callback and stop_callback():
-                    LOG.debug("Got stop signal")
-                    break
-
-                # skip self IP address
-                if victim.ip_addr in self._ip_addresses:
-                    continue
-
-                # skip IPs marked as blocked
-                if victim.ip_addr in WormConfiguration.blocked_ips:
-                    LOG.info("Skipping %s due to blacklist" % victim)
-                    continue
+            # skip self IP addresses
+            victim_chunk = [x for x in victim_chunk if x.ip_addr not in self._ip_addresses]
+            # skip IPs marked as blocked
 
-                LOG.debug("Scanning %r...", victim)
-                pingAlive = Pinger.is_host_alive(victim)
-                tcpAlive = TCPscan.is_host_alive(victim)
+            bad_victims = [x for x in victim_chunk if x.ip_addr in WormConfiguration.blocked_ips]
+            for victim in bad_victims:
+                LOG.info("Skipping %s due to blacklist" % victim)
+            victim_chunk = [x for x in victim_chunk if x.ip_addr not in WormConfiguration.blocked_ips]
```

Comment:
I don't like that we iterate over the same array 3 times. Create a method called smth like `should_scan_ip(ip)`. In that method check if ip is not self IP or blocked. Then do a single list comprehension with that function.

Reply:
Agreed, but I would like to keep the documentation on what victims are we skipping. Will refactor.